050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,718 @@
+
+### Goals
+
+* A unidirectional topic operator which is KRaft compatible
+* Support for multiple namespaces
+
+### Non-goals
+
+* Supporting RF change (this could be added later)
+
+### CR API
+
+We would remain compatible with the existing `KafkaTopic` resource at the API level. 
+This means that all existing resources could be handled by the new operator and all existing fields of the `status` would still be present for applications that consume the status.
+However, the semantics of a `KafkaTopic` would change incompatibly: The `spec` would no longer be updated when Kafka-side configuration changed.
+Instead if Kafka-side configuration was changed out-of-band (e.g. by using the `Admin` client, or Kafka shell scripts) those changes would eventually (depending on the timed reconciliation interval) be reverted.
+
+As context for the rest of this document, an example `KafkaTopic` looks like this:
+
+```yaml=
+kind: KafkaTopic
+metadata:
+  generation: 123
+  name: consumer.offsets
+spec:
+  topicName: __consumer_offsets
+  replicas: 3
+  partitions: 50
+  config:
+    retention.ms: 123456789
+status:
+  topicName: __consumer_offsets
+  observedGeneration: 123
+  conditions:
+  - type: Ready
+    status: True
+    lastTransitionTime: 20230301T103000Z
+```
+
+Where: 
+  * `spec.topicName` is optional and needs only be used for topic names which are not legal as kube resources. 
+    When this field is absent the name of the topic is Kafka is taken from the `metadata.name`.
+  * `status.topicName` is used to detect a change to the `spec.topicName`, which results in an error
+  * `metadata.generation == status.observedGeneration` implies that the operator has reconciled the latest change to the `spec`.
+    Whether that reconciliation was successful is indicated by the `Ready` condition having status `True`.
+
+Some new fields will be introduced in the following text.
+
+### The "single resource" principal
+
+To avoid ambiguity we will continue to require that a single `KafkaTopic` is used to manage a single topic in a single Kafka cluster.
+
+The fact that there are legal topic names (in Kafka) that are not legal resource names (in Kube) provided the motivation for supporting `spec.topicName`, so that a topic in Kafka could be managed by a `KafkaTopic` with a different `metadata.name`.
+As such it's not possible for the Kube apiserver to enforce the single resource principal because multiple `KafkaTopic` resource might refer to the same topic (i.e. the uniqueness constraint on `metadata.name` is insufficient).
+It is therefore left to the operator to detect if there are multiple `KafkaTopics` which point to the same topic in Kafka.
+
+To do this, the operator will keep an in-memory mapping of topic name to (namespace, name)-pairs.
+This will allow us to detect the case where multiple resources are attempting to manage the same topic in Kafka.
+When this happens both resources' `Ready` status will change to `False` with suitable `reason`.
```

Comment:
I guess we could do that. It's not impossible for the creationTimestamps to be the same, but very unlikely. I think this will work, but I'll confirm.